stl/inc/ranges

```diff
@@ -9553,6 +9553,595 @@ namespace ranges {
+            range_difference_t<_Maybe_const<_Const, _First>>, range_difference_t<_Maybe_const<_Const, _Rest>>...>;
+
+        template <bool _Const>
+        using _Size_type = common_type_t<_Optimal_size_type<_Const>, range_size_t<_Maybe_const<_Const, _First>>,
+            range_size_t<_Maybe_const<_Const, _Rest>>...>;
+
+        /* [[no_unique_address]] */ tuple<_First, _Rest...> _Bases;
+
+        template <bool _Const>
+        class _Iterator {
+        private:
+            friend cartesian_product_view;
+            using _Parent_t = _Maybe_const<_Const, cartesian_product_view>;
+
+            _Parent_t* _Parent = nullptr;
+            tuple<iterator_t<_Maybe_const<_Const, _First>>, iterator_t<_Maybe_const<_Const, _Rest>>...> _Current;
+
+            template <size_t _Index = sizeof...(_Rest)>
+            constexpr void _Next() {
+                auto& _It = _STD get<_Index>(_Current);
+                ++_It;
+                if constexpr (_Index > 0) {
+                    auto& _Range = _STD get<_Index>(_Parent->_Bases);
+                    if (_It == _RANGES end(_Range)) {
+                        _It = _RANGES begin(_Range);
+                        _Next<_Index - 1>();
+                    }
+                }
+            }
+
+            template <size_t _Index = sizeof...(_Rest)>
+            constexpr void _Prev() {
+                auto& _It = _STD get<_Index>(_Current);
+                if constexpr (_Index > 0) {
+                    auto& _Range = _STD get<_Index>(_Parent->_Bases);
+                    if (_It == _RANGES begin(_Range)) {
+                        _It = _Cartesian_common_arg_end(_Range);
+                        _Prev<_Index - 1>();
+                    }
+                }
+                --_It;
+            }
+
+            template <size_t... _Indices>
+            _NODISCARD constexpr bool _Entire_tail_at_begin(index_sequence<_Indices...>) const {
+                return ((_STD get<_Indices + 1>(_Current) == _RANGES begin(_STD get<_Indices + 1>(_Parent->_Bases)))
+                        && ...);
+            }
+
+            template <size_t _Index = sizeof...(_Rest)>
+            constexpr void _Advance(const _Difference_type<_Const> _Off) {
+                if (_Off == 0) {
+                    return;
+                }
+
+                auto& _Range = _STD get<_Index>(_Parent->_Bases);
+                auto& _It    = _STD get<_Index>(_Current);
+                using _Rng   = remove_reference_t<decltype(_Range)>;
+                using _Iter  = remove_reference_t<decltype(_It)>;
+
+                if constexpr (_Index > 0) {
+                    _STL_INTERNAL_STATIC_ASSERT(sized_range<_Rng>);
+                    const auto _Size       = static_cast<_Difference_type<_Const>>(_RANGES ssize(_Range));
+                    const auto _Begin      = _RANGES begin(_Range);
+                    const auto _It_off     = static_cast<_Difference_type<_Const>>(_It - _Begin);

```

Comment:
`_Range` is requires to be random-access here, since this is implementation of `operator+=`:
```c++
template <bool _Const, class _First, class... _Rest>
concept _Cartesian_product_is_random_access = (random_access_range<_Maybe_const<_Const, _First>> && ...
                                               && (random_access_range<_Maybe_const<_Const, _Rest>>
                                                   && sized_range<_Maybe_const<_Const, _Rest>>) );
```
I'm just going to remove this internal assertion, it is unnecessary.